### ISSUE TITLE
Make writing a single bam file in spark faster

### ISSUE THREAD

**Issue by droazen** [broadinstitute/gatk#1015]
`ReadsSparkSink.writeReadsSingle()` is unusably slow (even for small ouputs like a 200MB bam). We suspect this is due to its approach of calling `coalesce(1)`:

        final JavaPairRDD<GATKRead, SAMRecordWritable> out =
                rddSamRecordWriteable.sortByKey(new ReadCoordinateComparator(header)).coalesce(1);

We need to make it much faster in order to be able to use it in practice.

**Comment by droazen** [comment]
@tomwhite or @laserson, could one of you take this?

**Comment by tomwhite** [comment]
Sure. I think this is a dup of https://github.com/broadinstitute/gatk/issues/940

**Comment by droazen** [comment]
From @akiezun: "I tested on a 30Gb file, trying to write to HDFS. It wrote more than 1.4 TB to disk and took hours to complete."

**Comment by droazen** [comment]
@tomwhite Have you had any time to look into this one? This is a pretty high priority to fix for alpha

**Comment by laserson** [comment]
I'll try to take a look later this week as well.  Do you have an example "failing" command you could share for me to start with?

**Comment by droazen** [comment]
@akiezun Can you provide @laserson with a command that reproduces the issue?

**Comment by tomwhite** [comment]
Thanks for picking this up Uri. (BTW I'm out for the rest of this week and all of next.)

**Comment by laserson** [comment]
Wow.

```
build/install/gatk/bin/gatk PrintReadsSpark \
    --shardedOutput \
    --input src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam \
    --output tmp.bam

build/install/gatk/bin/gatk PrintReadsSpark \
    --input src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam \
    --output tmp.bam
```

What a difference...

**Comment by laserson** [comment]
Just to set some expectations, the *real* answer here is that for production pipelines, you will probably never want to write just a single output file.  Hadoop/Spark work well because of their distributed nature, and working on sharded data is a pretty crucial part of that.  In my mind, writing out single files is merely good manners.

With that in mind, I'm looking into how to make this less horrible.

**Comment by laserson** [comment]
Here are some initial thoughts, after working on my local machine.  It appears that writing a single BAM file is slowed significantly on the sort when the tasks have to spill to disk.  So:
* Lots of memory is good.  Decreases the chance of spilling.
* The sort itself should use multiple partitions.  This should decrease the amount of data going to each reducer/sort task.
* When we end up writing to a single partition, we actually want to induce an additional shuffle.  If we perform `.coalesce(1)` without the shuffle, then the `sortByKey` gets pushed into a single reducer task, ensuring lots of spilling and making things slower.

It might make sense to modify the API to look more like this:

`ReadsSparkSink.writeReads(..., boolean sortReads, int outputPartitions, int sortPartitions)`

This would enable a flag that decides whether to to a total sort of the data, and if so, let you set the number of sort tasks (perhaps we leave this one off).  Finally, the `outputPartitions` parameter lets us set how many partitions to perform on reshuffling.  If either are set to 0, they can just use default values.  If the output is set to 1 partition, it would write a single file.

**Comment by laserson** [comment]
Out of curiosity, what's the use case for writing to a single BAM file?

**Comment by droazen** [comment]
@laserson One use case would be that after running your Spark pipeline you wanted to run a file-based tool on the resulting reads. The vast majority of our tools are still file-based, and for tasks like performing QC on the output of a spark pipeline it's easiest to materialize a single file and run existing tools on it.

**Comment by laserson** [comment]
Yeah, this is definitely one of the biggest gotchas for people switching to Hadoop-style tools.  Though hopefully GATK4 will really help people move off of single files at least for large parts of their pipelines :)

**Comment by droazen** [comment]
We've just discovered that Hadoop BAM comes with a single BAM writer that is supposed to be fast. Let's try that one out, and put our efforts to develop our own on hold until we've evaluated the existing Hadoop BAM implementation.

**Comment by laserson** [comment]
sgtm.  Is someone else already implementing that or should I do it in this issue?

**Comment by tedsharpe** [comment]
Our current implementation of writeReadsSingle does a sortByKey, and then a coalesce(1).  From some hacking I've been doing, I think that it's actually sortByKey that's causing the big memory-spilling shuffle.  I'm guessing that this might be caused by the fact that our coordinate-sorted BAMs are not in Locatable order, so there's lots and lots of shuffling that has to happen.

**Comment by tomwhite** [comment]
@tedsharpe I agree with your analysis. I've created a PR to do the sorting in parallel, see https://github.com/broadinstitute/gatk/pull/1174.

**Comment by droazen** [comment]
@akiezun Can you evaluate whether https://github.com/broadinstitute/gatk/pull/1174 resolves this ticket, once it's merged? (assigning to you for that purpose)

**Comment by tomwhite** [comment]
My rough benchmarks while working on https://github.com/broadinstitute/gatk/pull/1174 indicated a substantial improvement:

> when I ran MD on a ~75MB BAM the runtime dropped from >30 mins to around 8 mins

So I think we can close this.

**Comment by akiezun** [comment]
@tomwhite before we close, how does this improved performance compare to non-spark writing of the same file (just use PrintReads vs PrintReadsSpark or something)?

**Comment by tomwhite** [comment]
@akiezun, I'm glad you asked me that, as I just found another significant improvement for Spark when writing a single BAM. See https://github.com/broadinstitute/gatk/pull/1190

I just ran on a ~75MB file and PrintReads takes ~12s, while PrintReadsSpark takes ~19s with this change.

**Comment by akiezun** [comment]
ha! this is awesome. does #1190 need to go in first?

**Comment by tomwhite** [comment]
Closing now #1190 is in.
